From a3abe1af9437aa4eadbab82dd837a43e09a89358 Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Wed, 22 Sep 2021 17:06:04 +0300
Subject: [PATCH] [SW-56687] gaudi2: arc: use arc viewpoint address to copy images

Currently, the logic used for converting 64 bit address received
as part of FIT loadaddr to ARC viewpoint address is incorrect which
causes hang while copying FIT images to loadaddr. This patch
fixes the logic and use correct ARC viewpoint address to copy
images.

Change-Id: Iebdaf86008f215bc09a97f6419c603509e671c6d
---

diff --git a/arch/gaudi2/arc/common.c b/arch/gaudi2/arc/common.c
index 61f3107..0ccbd1c 100644
--- a/arch/gaudi2/arc/common.c
+++ b/arch/gaudi2/arc/common.c
@@ -114,8 +114,12 @@
 
 void preboot_copy_image(u64 dst, const void *buf, int len, const char *arch)
 {
-	/* Prepare the ARC SRAM address */
-	u32 arc_dst = ((u32)dst & 0xFFFFFFF) | BIT(28);
+	u32 arc_dst;
+
+	if (!strcmp(arch, "arc"))
+		arc_dst = ((u32)dst & ARC_ADDR_MASK) | BIT(29);
+	else
+		arc_dst = ((u32)dst & ARM_ADDR_MASK) | BIT(28);
 
 	memcpy((void *)arc_dst, buf, len);
 }
diff --git a/arch/gaudi2/arc/common.h b/arch/gaudi2/arc/common.h
index 8aa4ce4..023b35a 100644
--- a/arch/gaudi2/arc/common.h
+++ b/arch/gaudi2/arc/common.h
@@ -21,6 +21,7 @@
 #ifdef TARGET_IS_ONIX
 #define SDRAM_BASE              0x10000000
 #define SDRAM_FULL_BASE         0x1001000200000000ull
+#define ARM_ADDR_MASK		0xFFFFFFFull
 #else
 /*
  * The SRAM region in ARC address extension is mapped to 0x10000000, but the
@@ -30,6 +31,7 @@
  */
 #define SDRAM_BASE              0x11000000
 #define SDRAM_FULL_BASE         0x1000FFFFFD000000ull
+#define ARM_ADDR_MASK		0x1FFFFFFull
 #endif
 
 #define GEN_REGION_BASE		0x20000000
@@ -76,6 +78,8 @@
 #define SPI_DMA_BAUDR_DEFAULT	0x2
 #define SPI_DMA_BAUDR_SRAM_DMA	0x4
 
+#define ARC_ADDR_MASK		0x7FFFFFFull
+
 #ifndef __ASSEMBLY__
 void init_secondary(void);
 extern void secondary_entry(void);
